treeherder/perf/autoclassify.py

```diff
@@ -0,0 +1,296 @@
+                    revisions[revision["revision"]] = []
+                revisions[revision["revision"]].append(alertsummary["id"])
+    return revisions
+
+
+def is_possible_upstream(downstream, upstream):
+    """Return a bool of whether the summary given could be upstream of the
+    given downstream
+    Args:
+        downstream - A downstream AlertSummary
+        upstream - A possible upstream AlertSummary
+    """
+    downstream_signatures = [alert["signature"] for alert in downstream['alerts']]
+    upstream_signatures = [alert["signature"] for alert in upstream["alerts"]]
+
+    return len(set(downstream_signatures).intersection(upstream_signatures)) > 0
+
+
+def identify_possible_upstreams(downstream, summaries):
+    """Return a list of AlertSummary that could be upstream
+    Args:
+        downstream - an AlertSummary that has been identified as downstream
+        summaries - a list of AlertSummary to filter from
+    """
+    return filter(lambda summary: is_possible_upstream(downstream, summary), summaries)
+
+
+def identify_true_upstream(downstream, upstreams, downstream_revisions, upstream_revision_map):
+    upstream_list = []
+    for revision in downstream_revisions[downstream["id"]]:
+        if revision not in upstream_revision_map.keys():
+            continue
+        for upstream in upstream_revision_map[revision]:
+            if upstream not in upstream_list:
+                upstream_list.append(upstream)
+    alerts = {}
+    for alert in downstream["alerts"]:
+        alerts[alert["id"]] = []
+        for upstream in upstreams:
+            if (alert["id"] in [upstream_alert["id"] for upstream_alert in upstream["alerts"]] and
+                upstream["id"] in upstream_list and
+                upstream["id"] not in alerts[alert["id"]]):
+                alerts[alert["id"]].append(upstream["id"])
+    return alerts
+
+
+def combine_upstreams(original, extra):
+    return original + extra
+
+
+def combine_upstream_revisions(original, extra):
+    revisions = {}
+    for key, value in original.iteritems():
+        revisions[key] = value
+
+    for key, value in extra.iteritems():
+        if key in revisions.keys():
+            revisions[key].extend(value)
+        else:
+            revisions[key] = value
+    return revisions
+
+
+def update_alert(alert_id, upstream_ids):
+    if len(upstream_ids) == 1:
+        alert = PerformanceAlert.objects.get(pk=alert_id)
+        alert.related_summary_id = [upstream_ids][0]
+        alert.save()
+
+
+def classify():
+    alertsummaries = get_all_alertsummaries()
+    revision_map = get_all_revision_map(alertsummaries)
+    downstreams = get_downstream_alertsummaries(alertsummaries, revision_map)
+    upstreams = get_upstream_alertsummaries(alertsummaries, downstreams)
+    upstream_revisions = get_upstream_revisions(upstreams, revision_map)
+    downstream_revisions = get_downstream_revisions(downstreams, revision_map)
```

Comment:
I think most of this shouldn't be necessary, since we should have direct access to the database. Just use the Django ORM to query for the data you need. Feel free to ask if you need help with this. :)